### PR TITLE
Improve balanceTx error if UTxOAssumptions are broken

### PR DESCRIPTION
**Pull request by Anviking**
- [ ] I have ...

### Issue Number

ADP-3056

### CODE REVIEW

File: lib/wallet/src/Cardano/Wallet/Write/Tx/Balance.hs [hunk 1]
```diff
@@ -277,6 +282,7 @@ data ErrSelectAssets
 
 data ErrBalanceTxInternalError
     = ErrUnderestimatedFee W.Coin SealedTx KeyWitnessCount
+    | ErrUTxOViolatesAssumptions UTxOAssumptionViolation
```
Comment (Anviking) [inline lib/wallet/src/Cardano/Wallet/Write/Tx/Balance.hs:285]: This is not actually an internal balanceTx error. In the HTTP API it is an internal _wallet_ error, but not here, only looking at balanceTx. Will move.

**Comment by Anviking** [comment]
Failing with:

```
test/unit/Cardano/Wallet/Api/TypesSpec.hs:978:9:
--
  | cardano-wallet-test-unit>   1) Cardano.Wallet.Api.Types, Api Errors, Every ApiErrorInfo constructor has a corresponding schema type
  | cardano-wallet-test-unit>        Falsified (after 1 test):
  | cardano-wallet-test-unit>          Missing ApiErrorInfo constructors for:
  | cardano-wallet-test-unit>          ["UTxOAssumptionsViolated"]
  | cardano-wallet-test-unit>          Each of these need a corresponding swagger type of the form:
  | cardano-wallet-test-unit>          x-errConstructorName
```

Not sure it makes full sense to have to provide user-facing specifications for all the err500 errors 🤔

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs [hunk 2]
```diff
@@ -3041,7 +3099,8 @@ instance Arbitrary Wallet' where
         genShelleyVkAddr = Cardano.shelleyAddressInEra
             <$> (Cardano.makeShelleyAddress
                 <$> genNetworkId
-                <*> genPaymentCredential -- only vk credentials
```
Comment (Anviking) [inline lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:3044]: Was also generating script credentials

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs [hunk 3]
```diff
@@ -2515,6 +2557,7 @@ balanceTransactionSpec = describe "balanceTransaction" $ do
 
     wallet = mkTestWallet (utxo [Coin 5_000_000])
 
+
```
Comment (Anviking) [inline lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:2560]:
```suggestion
```

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs [hunk 4]
```diff
@@ -2537,6 +2580,17 @@ balanceTransactionSpec = describe "balanceTransaction" $ do
...
+        AnyChangeAddressGenWithState g s0 ->
+            fst $ getChangeAddressGen g s0
+
+    byronUtxo coins = UTxO $ Map.fromList $ zip ins outs
+      where
+        ins = map (TxIn dummyHash) [0..]
+        outs = map (TxOut dummyBootstrapAddr . TokenBundle.fromCoin) coins
+        dummyHash = Hash $ B8.replicate 32 '0'
+
+
```
Comment (Anviking) [inline lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:2593]:
```suggestion
```

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs [hunk 5]
```diff
@@ -488,6 +489,8 @@ import qualified Cardano.Wallet.Primitive.Types.Tx.TxOut.Gen as TxOutGen
 import qualified Cardano.Wallet.Shelley.Compatibility as Compatibility
 import qualified Cardano.Wallet.Write.ProtocolParameters as Write
 import qualified Cardano.Wallet.Write.Tx as Write
+import Cardano.Wallet.Write.UTxOAssumptions
+    ( UTxOAssumptionViolation (UTxOAssumptionViolation) )
```
Comment (Anviking) [inline lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:493]: wrong section

**Comment by abailly** [comment]
Is this still relevant?

**Comment by Anviking** [comment]
Closing for now, but this is a fairly small task that would be nice to get back to at some point, for the sake of integrity of the `cardano-balance-tx` library and function. Without this change, misconfiguration of `balanceTx` may yield e.g. `ErrUnderestimatedFee` errors instead of `UTxOAssumptionViolation`.

As to whether the misconfiguration should be possible in the first place, I think it is reasonable, but that's another topic.

**Pull request closed without merging**